service/tbc/tbc.go

```diff
@@ -1110,11 +854,47 @@ func (s *Server) syncBlocks(ctx context.Context) {
 		go func() {
 			if err = s.SyncIndexersToBest(ctx); err != nil && !errors.Is(err, ErrAlreadyIndexing) && !errors.Is(err, context.Canceled) {
 				// XXX this is probably not a panic.
-				panic(fmt.Errorf("sync blocks: %w", err))
+				panic(fmt.Errorf("sync blocks: %T %w", err, err))
 			}
 
-			// Get block headers
-			s.pm.All(ctx, s.headersPeer)
+			// Get block headers that we missed during indexing.
+			if s.Synced(ctx).Synced {
+				s.mtx.Lock()
+				ib := make([]*chainhash.Hash, 0, len(s.invBlocks))
+				for k := range s.invBlocks {
+					ib = append(ib, &k)
+				}
+				clear(s.invBlocks)
+				s.mtx.Unlock()
+
+				// Fixup ib array to not ask for block headers
+				// we already have.
+				ib = slices.DeleteFunc(ib, func(h *chainhash.Hash) bool {
+					_, _, err := s.BlockHeaderByHash(ctx, h)
+					return err == nil
+				})
+
+				// Flush out blocks we saw during quiece.
+				log.Debugf("flush missed headers %v", len(ib))
+
+				if len(ib) == 0 {
+					log.Debugf("nothing to do")
+					return
+				}
+
+				hp := func(ctx context.Context, p *peer) {
+					if err = s.getHeadersByHashes(ctx, p, ib...); err != nil {
+						log.Errorf("missed block headers: %v %v",
+							p, err)
+						return
+					}
+				}
+				s.pm.All(ctx, hp)
+
+			} else {
+				log.Debugf("handle all")
+				s.pm.All(ctx, s.headersPeer)
+			}
```

Comment:
nit: I prefer avoiding `else`; so this is completely optional and there may be a reason you have avoided this which is not immediately obvious:
```suggestion
				return
			}

			log.Debugf("handle all")
			s.pm.All(ctx, s.headersPeer)
```

Reply:
This is one is deliberate to make it obvious since the if clause is lengthy. I am keeping it.